test-data/unit/check-classes.test

```diff
@@ -1702,6 +1729,77 @@ class C:
 tmp/foo.pyi:3: error: Invalid signature "def (foo.B) -> foo.A"
 tmp/foo.pyi:5: error: Invalid signature "def (foo.C, Any, Any) -> builtins.int"
 
+[case testReverseOperatorOrderingCase1]
+class A:
+    def __radd__(self, other: 'A') -> int: ...
+
+# Note: Python only tries calling __add__ and never __radd__, even though it's present
+A() + A()  # E: Unsupported left operand type for + ("A") \
+           # N: __radd__ will not be called when running 'A + A': must define __add__
+
+[case testReverseOperatorOrderingCase2]
+class A:
+    def __lt__(self, other: object) -> bool: ...
+
+# Not all operators have the above shortcut though.
+reveal_type(A() > A())   # E: Revealed type is 'builtins.bool'
+reveal_type(A() < A())   # E: Revealed type is 'builtins.bool'
+[builtins fixtures/bool.pyi]
+
+[case testReverseOperatorOrderingCase3]
+class A:
+    def __add__(self, other: B) -> int: ...
+
+class B:
+    def __radd__(self, other: A) -> str: ...  # E: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping
+
+# Normally, we try calling __add__ before __radd__
+reveal_type(A() + B())  # E: Revealed type is 'builtins.int'
+
+[case testReverseOperatorOrderingCase4]
+class A:
+    def __add__(self, other: B) -> int: ...
+
+class B(A):
+    def __radd__(self, other: A) -> str: ...  # E: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping
+
+# However, if B is a subtype of A, we try calling __radd__ first.
+reveal_type(A() + B())  # E: Revealed type is 'builtins.str'
+
+[case testReverseOperatorOrderingCase5]
+class A:
+    def __add__(self, other: B) -> int: ...
+    def __radd__(self, other: A) -> str: ...
```

Comment:
Just do double check: this is not an unsafe overlap because only one of the two can be ever called? Could you please add a clarifying comment here?

Reply:
Yep. I added a comment.